### PR TITLE
Enable calling Java methods that take arrays

### PR DESCRIPTION
**Pull request by paulproteus**
## PR Checklist:
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: rubicon/java/api.py [hunk 1]
```diff
@@ -170,6 +184,15 @@ def select_polymorph(polymorphs, args):
...
                 ])
+            elif isinstance(arg, Sequence) and len(arg) > 0:
+                # If arg is an iterable of all the same basic numeric type, then
+                # an array of that Java type can work.
+                if isinstance(arg[0], (int, jint)):
+                    if all((isinstance(item, (int, jint)) for item in arg)):
+                        arg_types.append([b'[I'])
+                elif isinstance(arg[0], (float, jfloat, jdouble)):
+                    if all((isinstance(item, (float, jfloat, jdouble)) for item in arg)):
+                        arg_types.append([b'[D', b'[F'])
```
Comment (freakboy3742) [inline rubicon/java/api.py:194]: What's the fallback case here? If `arg` is a sequence of non-zero length, and types *aren't* consistent, this won't throw an unknown argument error (or any other type of error) - it will move on to trying to create an option

File: rubicon/java/api.py [hunk 2]
```diff
@@ -104,6 +105,19 @@ def convert_args(args, type_names):
...
+                java.SetIntArrayRegion(jarg, 0, len(arg), (jint * len(arg))(*arg))
+                converted.append(jarg)
+            elif type_name[1] == ord(b'F'):
+                jarg = java.NewFloatArray(len(arg))
+                java.SetFloatArrayRegion(jarg, 0, len(arg), (jfloat * len(arg))(*arg))
+                converted.append(jarg)
+            elif type_name[1] == ord(b'D'):
+                jarg = java.NewDoubleArray(len(arg))
+                java.SetDoubleArrayRegion(jarg, 0, len(arg), (jdouble * len(arg))(*arg))
+                converted.append(jarg)
```
Comment (freakboy3742) [inline rubicon/java/api.py]: Is there any reason to not include byte, char, bool, and any other primitive for which there is a SetXArrayRegion method?

File: rubicon/java/types.py [hunk 3]
```diff
@@ -19,7 +19,7 @@
 jbyte = c_byte
 jchar = c_wchar
 jshort = c_short
-jint = c_int
+jint = c_uint32
```
Comment (freakboy3742) [inline rubicon/java/types.py:22]: Are you sure about this? Aren't Java ints 32 bit signed?

File: rubicon/java/api.py (hunk 1 above)
Comment (freakboy3742) [reply rubicon/java/api.py:194]: This would also seem to rule out passing in an empty array under any circumstances...

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Basic mechanics look sound and well tested. The edge cases are my only concern - both in terms of the ways the API can be used, and APIs that people might want to use. Details inline.

File: rubicon/java/api.py (hunk 2 above)
Comment (paulproteus) [reply rubicon/java/api.py]: Bytes are already smoothly handled by the `b'xyz'` handling a few lines before this.

`char` and `bool`... sure... but I just haven't really seen 'em in the wild, so I'm just being lazy.

I can try adding them.

File: rubicon/java/api.py (hunk 1 above)
Comment (paulproteus) [reply rubicon/java/api.py:196]: Both should be clarified in a fresh push

File: rubicon/java/types.py (hunk 3 above)
Comment (paulproteus) [reply rubicon/java/types.py]: Right-o -- I meant to specify c_int32. Fixed now.

I am a little weirded-out to say `c_int` rather than `c_int32`, so I like the latter, but I'm OK if we don't make that change.

File: rubicon/java/api.py (hunk 2 above)
Comment (paulproteus) [reply rubicon/java/api.py]: For `char` arrays, I want to punt. For the rest, they're now handled.

File: tests/test_rubicon.py [hunk 4]
```diff
@@ -238,6 +238,49 @@ def test_polymorphic_static_method(self):
...
+
+    def test_heterogenous_list(self):
+        """A list of mixed types raise an exception when trying to find the right Java method."""
+        Example = JavaClass("org/beeware/rubicon/test/Example")
+        with self.assertRaises(ValueError):
+            Example.sum_all_ints([1, "two"])
+        with self.assertRaises(ValueError):
+            Example.sum_all_floats([1.0, "two"])
+        with self.assertRaises(ValueError):
+            Example.sum_all_doubles([1.0, "two"])
```
Comment (freakboy3742) [inline tests/test_rubicon.py:254]: Missing test case here is 'first argument == "two"'

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Almost there - one more edge case I can see.

File: rubicon/java/api.py [hunk 5]
```diff
@@ -170,6 +187,27 @@ def select_polymorph(polymorphs, args):
...
+                elif isinstance(arg[0], (int, jint)):
+                    if all((isinstance(item, (int, jint)) for item in arg)):
+                        arg_types.append([b'[I'])
+                    else:
+                        raise ValueError("Unable to treat all data in list as integers")
+                elif isinstance(arg[0], (float, jfloat, jdouble)):
+                    if all((isinstance(item, (float, jfloat, jdouble)) for item in arg)):
+                        arg_types.append([b'[D', b'[F'])
+                    else:
+                        raise ValueError("Unable to treat all data in list as floats/doubles")
```
[the remark this answers is not in the document]
Comment (paulproteus) [reply rubicon/java/api.py:210]: Added

File: tests/test_rubicon.py (hunk 4 above)
Comment (paulproteus) [reply tests/test_rubicon.py:256]: Added

**Review by freakboy3742 — APPROVED** [review]
Looks good to me!
